src/main/java/seedu/address/commons/util/timetable/TimetableParserUtil.java

```diff
@@ -0,0 +1,217 @@
+            }
+
+            return longUrl;
+        } catch (IOException ioe) {
+            throw new ParseException(MESSAGE_INVALID_URL);
+        }
+    }
+
+    /**
+     * Parses the full NUSMods link into TimetableData format
+     * @param url Full NUSMods link
+     * @return TimetableData parsed from link
+     * @throws ParseException
+     */
+    public static TimetableData parseLongUrl(String url) throws ParseException {
+
+        String[] urlParts = url.split("/");
+        ArrayList<Lesson> totalLessonList = new ArrayList<Lesson>();
+        ArrayList<Lesson> lessonsToAddFromModule;
+
+        String semNum = urlParts[4];
+        String[] toParse = urlParts[5].split("\\?");
+        String[] modules = toParse[1].split("&");
+
+        for (String module: modules) {
+            lessonsToAddFromModule = parseModule(module, semNum);
+            for (Lesson lessonToAdd: lessonsToAddFromModule) {
+                totalLessonList.add(lessonToAdd);
+            }
+        }
+        return new TimetableData(totalLessonList);
+    }
+
+    /**
+     *
+     * @param module String that contains module code and lessons selected
+     * @param semNum Semester number
+     * @return ArrayList of lessons taken
+     * @throws ParseException when invalid lessons are received
+     */
+    public static ArrayList<Lesson> parseModule(String module, String semNum) throws ParseException {
+        ArrayList<Lesson> lessonListFromApi;
+        ArrayList<Lesson> lessonsTakenList;
+        lessonsTakenList = new ArrayList<Lesson>();
+        String[] moduleInfo = module.split("=");
+        String[] semNumToParse = semNum.split("-");
+        int semNumInt = Integer.parseInt(semNumToParse[1]);
+        String moduleCode = moduleInfo[0];
+        String[] lessonsTaken = moduleInfo[1].split(",");
+
+        lessonListFromApi = obtainModuleInfoFromApi(moduleCode, semNumInt);
+
+        for (String lessonTaken : lessonsTaken) {
+            String[] lessonToParse = lessonTaken.split(":");
+            String lessonType = convertShortFormToLong(lessonToParse[0]);
+            String lessonNum = lessonToParse[1];
+
+            for (Lesson lessonFromApi: lessonListFromApi) {
+                if (lessonType.equalsIgnoreCase(lessonFromApi.getLessonType())
+                        && lessonNum.equalsIgnoreCase(lessonFromApi.getClassNo())) {
+                    lessonsTakenList.add(lessonFromApi);
+                }
+            }
+        }
+
+        return lessonsTakenList;
+    }
+
+    /**
+     * Access NUSMods API and obtains and parses the json file
+     * @param moduleCode
+     * @param semNum current semester number
+     * @return Total list of lessons a module has
+     * @throws ParseException
+     */
+    public static ArrayList<Lesson> obtainModuleInfoFromApi(String moduleCode, int semNum) throws ParseException {
+        LocalDate currentDate = LocalDate.now();
+        String acadYear;
+
+        // Calculate current academic year
```

Comment:
Can be improved to apply SLAP, create another function to calculate current academic year